### PR TITLE
Texture Abstraction Class

### PR DESCRIPTION
**Pull request by Burke-Daniel**
This PR adds a Texture class, which uses stbi to abstract away the process of creating textures. Up to a minimum of 16 textures can be created, with the ability to set the texture wrapping and texture filtering options, and to bind and unbind individual textures. The sandbox example has also been modified to render the container image from the learnOpenGL tutorial to the window.

### CODE REVIEW

File: Aquarius/Src/Aquarius/Renderer/Texture.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		generateTexture(texture, hasAlpha);
+	}
+
+	bool Texture::generateTexture(const std::string& texture, bool hasAlpha)
+	{
+		bool success = false;
+
+		glGenTextures(1, &m_ID);
+		glBindTexture(GL_TEXTURE_2D, m_ID);
+		m_TextureUnitNum = GL_TEXTURE0 + s_NextTextureUnitNum++;
```
Comment (colton-smith) [inline Aquarius/Src/Aquarius/Renderer/Texture.cpp]: Nice

File: Aquarius/Src/Aquarius/Renderer/Texture.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, static_cast<int>(m_Configuration.textureWrapS));
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, static_cast<int>(m_Configuration.textureWrapT));
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, static_cast<int>(m_Configuration.minTextureFiltering));
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, static_cast<int>(m_Configuration.magTextureFiltering));
+		unsigned char* data = stbi_load(texture.c_str(), &m_Width, &m_Height, &m_NumColorChannels, 0);
```
Comment (colton-smith) [inline Aquarius/Src/Aquarius/Renderer/Texture.cpp:29]: Prior to calling this, the following should be called somewhere, once:
```
 stbi_set_flip_vertically_on_load(true);
 ```

File: Sandbox/Shaders/fragmentShader.glsl [hunk 3]
```diff
@@ -3,8 +3,11 @@
...
 in vec3 fColor;
+in vec2 fTexCoord;
+
+uniform sampler2D fTexture;
 
 void main()
 {
-     FragColor = vec4(fColor, 1.0);
+     FragColor = texture(fTexture, fTexCoord) * vec4(fColor, 1.0); 
 }
```
Comment (colton-smith) [inline Sandbox/Shaders/fragmentShader.glsl]: Is it cool if we make another set of vertex/frag shaders? I think we should have one that is pliable and then one thats static and used in the quadRenderer

Comment (colton-smith) [reply Sandbox/Shaders/fragmentShader.glsl]: For the quad renderer, I'm thinking the texture coords will just be uniforms and not vertex data

File: Aquarius/Src/Aquarius/Renderer/Texture.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		
+		if (data != nullptr)
+		{
+			int format = hasAlpha ? GL_RGBA : GL_RGB;
+
```
Comment (colton-smith) [inline Aquarius/Src/Aquarius/Renderer/Texture.cpp]: Might be better to use uint32_t here for consistency

Comment (colton-smith) [reply Aquarius/Src/Aquarius/Renderer/Texture.cpp:31]: To avoid narrowing conversions, GLenum is 32 bits

**Review by colton-smith — CHANGES_REQUESTED** [review]
Apart from a couple of small things I noted in the comments:

What happens if a texture object is destroyed? I think that textures unit number will be "lost" so to speak. 

 Might be nice to somehow separate texture asset creation from managing the texture slots

File: Sandbox/Shaders/fragmentShader.glsl (hunk 3 above)
Comment (colton-smith) [reply Sandbox/Shaders/fragmentShader.glsl]: Also, I was wondering if the quad renderer shader should be moved in-engine

Comment (Burke-Daniel) [reply Sandbox/Shaders/fragmentShader.glsl]: Added default and modifiable shaders to the Shaders folder in Sandbox, I'll make a comment on your PR to move your changes to the shader for quad rendering to within the engine

File: Aquarius/Src/Aquarius/Renderer/Texture.cpp (hunk 4 above)
Comment (Burke-Daniel) [reply Aquarius/Src/Aquarius/Renderer/Texture.cpp]: Good call

**Comment by Burke-Daniel** [comment]
Just added a couple of commits addressing your comments, I'll get you to take another look @colton-smith

File: Aquarius/Src/Aquarius/Renderer/Texture.cpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+		stbi_image_free(data);
+		glBindTexture(GL_TEXTURE_2D, 0);
+
+		return success;
+	}
+
+	void Texture::bind(uint32_t textureSlot) const
+	{
+		glActiveTexture(textureSlot);
```
Comment (colton-smith) [inline Aquarius/Src/Aquarius/Renderer/Texture.cpp]: Maybe this should be like, GL_TEXTURE0 + textureSlot?

File: Aquarius/Src/Aquarius/Renderer/Texture.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+		int getHeight() const { return m_Height; }
+		int getNumColorChannels() const { return m_NumColorChannels; }
+
+		void setConfiguration(const TextureConfiguration& configuration) { m_Configuration = configuration; }
+
+	private:
+		uint32_t m_ID = 0;
+		int m_Width = 0;
+		int m_Height = 0;
+		int m_NumColorChannels = 0;
```
Comment (colton-smith) [inline Aquarius/Src/Aquarius/Renderer/Texture.h:56]: I dont think this is ever set ?

File: Sandbox/Sandbox.cpp [hunk 7]
```diff
@@ -16,24 +16,31 @@ void SandboxLayer::onCreation()
...
+	m_texture = std::make_shared<Aquarius::Texture>("Assets/container.jpg");
+	m_texture->bind(GL_TEXTURE0);
 }
 
 void SandboxLayer::onUpdate(Aquarius::timeDelta_t)
 {
 	Aquarius::Renderer::ClearColor({ 0.2, 0.3, 0.7 });
 	Aquarius::Renderer::Clear();
+
+	m_texture->bind(GL_TEXTURE0);
```
Comment (colton-smith) [inline Sandbox/Sandbox.cpp]: I think it makes sense to pass this an offset from GL_TEXTURE0

**Review by colton-smith — CHANGES_REQUESTED** [review]
Couple of small things then we are GTM this!

- color channel member is never set
- We should just provide an int offset to bind, then bind will just use that as an offset from GL_TEXTURE0 -> this way calling code doesn't need the GL_TEXTURE enum.

File: Aquarius/Src/Aquarius/Renderer/Texture.h (hunk 6 above)
Comment (Burke-Daniel) [reply Aquarius/Src/Aquarius/Renderer/Texture.h:56]: It gets set by the call to `stbi_load` in `generateTexture`

Comment (colton-smith) [reply Aquarius/Src/Aquarius/Renderer/Texture.h:56]: Jesus i commented on that previously as well

**Review by colton-smith — APPROVED** [review]
GTM

File: Aquarius/Src/Aquarius/Renderer/Texture.cpp [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, static_cast<int>(m_Configuration.textureWrapT));
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, static_cast<int>(m_Configuration.minTextureFiltering));
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, static_cast<int>(m_Configuration.magTextureFiltering));
+		unsigned char* data = stbi_load(texture.c_str(), &m_Width, &m_Height, &m_NumColorChannels, 0);
+		
+		if (data != nullptr)
+		{
+			uint32_t format = hasAlpha ? GL_RGBA : GL_RGB;
+
+			glTexImage2D(GL_TEXTURE_2D, 0, format, m_Width, m_Height, 0, format, GL_UNSIGNED_BYTE, data);
```
Comment (colton-smith) [inline Aquarius/Src/Aquarius/Renderer/Texture.cpp]: Technically the first format argument is the internal format and should be RGBA8 or RGB8 (RGBA works only for backwards compatibility according to youtube)

**Review by colton-smith — CHANGES_REQUESTED** [review]
- Destructor should call glDeleteTextures
- 2nd argument of glTexImage2D is the internal format and should have a number attached which is the bits per pixel

File: Aquarius/Src/Aquarius/Renderer/Texture.cpp (hunk 8 above)
Comment (Burke-Daniel) [reply Aquarius/Src/Aquarius/Renderer/Texture.cpp]: Ahh I knew they were different but in learnopenGL they used GL_RGB and RGBA for both, but this is an easy fix

**Review by colton-smith — APPROVED** [review]
send it

**Pull request merged**